### PR TITLE
Command and basic UI for custom presence status messages.

### PR DESCRIPTION
**Pull request by Half-Shot**
To make a start on https://github.com/vector-im/vector-web/issues/1528, I've added the command /status to set yourself to one of the 3 states and to leave a custom message. This will only work if you are on the latest synapse develop right now due to a bug. I've left offline and unavaliable in there for the sake of having it there, but I believe synapse will always set you to online if you have any activity.

In addition, the MemberTiles will now display status_msg if it has been set alongside your presence state.

~~This requires https://github.com/matrix-org/matrix-js-sdk/pull/167 in order to work. *~~ Which has been merged :)

Signed-off-by: Will Hunt will@half-shot.uk

----

Fixes https://github.com/vector-im/riot-web/issues/478
~ Travis

### CODE REVIEW

File: src/SlashCommands.js [hunk 1]
```diff
@@ -271,7 +271,22 @@ var commands = {
             }
         }
         return reject(this.getUsage());
-    })
+    }),
+    status: new Command("status", "<online|offline|unavailable> [<status message>]", function(roomId, args) {
```
Comment (richvdh) [inline src/SlashCommands.js]: could do with an empty line here

File: src/SlashCommands.js [hunk 2]
```diff
@@ -271,7 +271,22 @@ var commands = {
...
+      if (args) {
+          let matches = args.match(/^(online|offline|unavailable)( +(.*))?$/);
+          if (matches) {
+              let opt = { presence: matches[1] };
+
+              if (matches[2]) {
+                opt.status_msg = matches[2];
```
Comment (richvdh) [inline src/SlashCommands.js]: don't you want `matches[3]`?

File: src/SlashCommands.js [hunk 3]
```diff
@@ -271,7 +271,22 @@ var commands = {
             }
         }
         return reject(this.getUsage());
-    })
+    }),
+    status: new Command("status", "<online|offline|unavailable> [<status message>]", function(roomId, args) {
+      if (args) {
+          let matches = args.match(/^(online|offline|unavailable)( +(.*))?$/);
```
Comment (richvdh) [inline src/SlashCommands.js]: perhaps you want `([^ ].*)` for the innermost pattern, to make sure that the status message is non-empty.

File: src/components/views/rooms/EntityTile.js [hunk 4]
```diff
@@ -60,14 +60,16 @@ module.exports = React.createClass({
         showInviteButton: React.PropTypes.bool,
         shouldComponentUpdate: React.PropTypes.func,
         onClick: React.PropTypes.func,
-        suppressOnHover: React.PropTypes.bool
+        suppressOnHover: React.PropTypes.bool,
+        presenceStatus: React.PropTypes.string,
```
Comment (richvdh) [inline src/components/views/rooms/EntityTile.js]: why not put this next to the other `presence` entries in the list?

File: src/components/views/rooms/PresenceLabel.js [hunk 5]
```diff
@@ -34,7 +34,10 @@ module.exports = React.createClass({
         currentlyActive: React.PropTypes.bool,
 
         // offline, online, etc
-        presenceState: React.PropTypes.string
+        presenceState: React.PropTypes.string,
+
+        // status_msg
+        presenceStatus: React.PropTypes.string
```
Comment (richvdh) [inline src/components/views/rooms/PresenceLabel.js]: please include a trailing comma

File: src/SlashCommands.js (hunk 2 above)
Comment (Half-Shot) [inline src/SlashCommands.js]: The pattern only returns 3 matches, and I'd want the last one?

**Comment by richvdh** [comment]
ok to test

**Comment by richvdh** [comment]
looks better. @ara4n: any comments on the UX?

**Comment by Half-Shot** [comment]
Bump?

**Comment by ara4n** [comment]
uh, i seem to have accidentally managed to miss this for months; sorry :(

My concern here is whether the idea of custom presence states is actually how we want to do this in terms of the spec.  It feels like the current online/idle/offline states have pretty well defined semantics, and are basically an enum (which may be internationalised etc. in future).

Going and then using that field as a free-form string feels pretty wrong.  Instead it feels like freeform status messages should be an orthogonal separate profile field of some kind.

I am painfully aware that the spec says you can put custom statuses in presence, but I think this is probably a mistake that wasn't fully thought through at the time.

@erikjohnston do you have thoughts on this (given you've been playing with both presence & profiles recently?)

**Comment by t3chguy** [comment]
Is this made redundant by @turt2live's presence stuffs?

**Comment by turt2live** [comment]
no

**Comment by t3chguy** [comment]
oh it has custom status stuff too, nvm

**Comment by Half-Shot** [comment]
I just remembered about this PR again and I'm wondering if it's worth resurrecting this on the basis that we still don't have a spec solution, and it's in use by bridges in the wild.

**Comment by Half-Shot** [comment]
@ara4n >does this just do arbitrary messages for custom presence, or is it meant to be arbitrary states like "away" or "asleep"?

It's custom messages because I feel like if we want to restrict the messages to a subset for uniformity, it should be specced.

**Comment by turt2live** [comment]
This fixes https://github.com/vector-im/riot-web/issues/478

**Comment by Half-Shot** [comment]
@turt2live is this useful to you at all, if I were to clean it up and look at getting it merged?

**Comment by turt2live** [comment]
Not for my particular use case, sadly. It'd be good to see it landed independently, although I think presence is scheduled for a rethink.

**Comment by Half-Shot** [comment]
>although I think presence is scheduled for a rethink

Which was the same reason this kinda stalled last time :/

Anyway, I've fixed this up as best I can without testing it.

**Comment by turt2live** [comment]
> Which was the same reason this kinda stalled last time :/

The eternal story of presence :(

**Comment by jryans** [comment]
@Half-Shot Would it make sense to close this then, since the path to merging seems unclear at the moment? We can always re-open / recover the code in the future.

**Comment by Half-Shot** [comment]
@jryans Yes, although I'd like an issue opened to state that status_msg and the new experimental type are at odds in terms of UX (even if not the same thing technically).

**Comment by jryans** [comment]
I filed https://github.com/vector-im/riot-web/issues/8029 to capture the UX conflicts between these two related features. Please edit or comment on the new issue as needed.

**Pull request closed without merging**